### PR TITLE
#164 ConnectionListener#onDisconnected_is_called_twice_for_every_disconnect

### LINKED ISSUE #164: ConnectionListener#onDisconnected is called twice for every disconnect
(when using the UberHandler at least)

It's called here

https://github.com/OpenHFT/Chronicle-Network/blob/5d53d322e4b1713c6e8ce24326038f4e382a0b53/src/main/java/net/openhft/chronicle/network/cluster/handlers/UberHandler.java#L157

and here

https://github.com/OpenHFT/Chronicle-Network/blob/5d53d322e4b1713c6e8ce24326038f4e382a0b53/src/main/java/net/openhft/chronicle/network/cluster/handlers/HeartbeatHandler.java#L227-L228

### PR DESCRIPTION
**Pull request by jkolobok**

### CODE REVIEW

**Review by JerryShea — APPROVED** [review]

**Review by JerryShea — APPROVED** [review]

**Review by nicktindall — CHANGES_REQUESTED** [review]
Don't merge this yet, need to add a test for a Heartbeat handler timeout. The current test only covers a TCP handler timeout.

I will add it tomorrow.

File: src/main/java/net/openhft/chronicle/network/cluster/handlers/HeartbeatHandler.java [hunk 1]
```diff
@@ -134,12 +136,20 @@ public void onRead(@NotNull WireIn inWire, @NotNull WireOut outWire) {
...
+                } catch (Exception e) {
+                    Jvm.error().on(getClass(), "close:", e);
+                    throw Jvm.rethrow(e);
+                }
+            }
+            lastTimeMessageReceived = Long.MAX_VALUE;
+            Closeable closable = closable();
+            if (closable != null && !closable.isClosed()) {
+                Closeable.closeQuietly(closable);
+            }
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/network/cluster/handlers/HeartbeatHandler.java:151]: If we make this close idempotent, and we call onDisconnected in the close method, we only end up calling it once (it always gets closed when disconnecting)

File: src/main/java/net/openhft/chronicle/network/cluster/handlers/HeartbeatHandler.java [hunk 2]
```diff
@@ -224,9 +234,6 @@ public boolean action() throws InvalidEventHandlerException {
 
             if (hasHeartbeats != prev) {
                 if (!hasHeartbeats) {
-                    connectionMonitor.onDisconnected(HeartbeatHandler.this.localIdentifier(),
-                            HeartbeatHandler.this.remoteIdentifier(), HeartbeatHandler.this.nc().isAcceptor());
-
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/network/cluster/handlers/HeartbeatHandler.java:229]: Don't need to call it here, it's called in `HeartbeatHandler.this.close()`

File: src/main/java/net/openhft/chronicle/network/cluster/handlers/UberHandler.java [hunk 3]
```diff
@@ -299,4 +286,4 @@ public String toString() {
                 ", localIdentifier=" + localIdentifier +
                 '}';
     }
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/network/cluster/handlers/UberHandler.java:284]: Only reset heartbeat when we RECEIVE a message, not when we send one (we send regular heartbeats that will suppress a timeout if the other side stops sending heartbeats)

**Review by nicktindall — APPROVED** [review]

**Pull request merged**